cirq/circuits/circuit.py

```diff
@@ -407,6 +408,208 @@ def _prev_moment_available(
                 last_available = k
         return last_available
 
+    def reachable_frontier_from(
+            self,
+            start_frontier: Dict[ops.QubitId, int],
+            *,
+            is_blocker: Callable[[ops.Operation], bool] = lambda op: False
+    ) -> Dict[ops.QubitId, int]:
+        """Determines how far can be reached into a circuit under certain rules.
+
+        The location L = (qubit, moment_index) is *reachable* if and only if:
+
+            a) L is one of the items in `start_frontier`.
+
+            OR
+
+            b) There is no operation at L and prev(L) = (qubit, moment_index-1)
+                is reachable and L is within the bounds of the circuit.
+
+            OR
+
+            c) There is an operation P covering L and, for every location
+                M = (q', moment_index) that P covers, the location
+                prev(M) = (q', moment_index-1) is reachable. Also, P must not be
+                classified as a blocker by the given `is_blocker` argument.
+
+        In other words, the reachable region extends horizontally along each
```

Comment:
horizontally is a bit awkward since sometimes we draw the diagrams vertically.  Think also you want to say forward in time.